core/trino-main/src/main/java/io/trino/sql/planner/LogicalPlanner.java

```diff
@@ -451,29 +452,38 @@ private RelationPlan getInsertPlan(
                 continue;
             }
             Symbol output = symbolAllocator.newSymbol(column.getName(), column.getType());
+            Expression expression;
+            Type tableType = column.getType();
+            boolean missingValueForNotNullableColumn = false;
             int index = insertColumns.indexOf(columns.get(column.getName()));
             if (index < 0) {
                 if (supportsMissingColumnsOnInsert) {
                     continue;
                 }
-                Expression cast = new Cast(new NullLiteral(), toSqlType(column.getType()));
-                assignments.put(output, cast);
-                insertedColumnsBuilder.add(column);
+                if (!column.isNullable()) {
+                    expression = createNullNotAllowedFailExpression(column.getName(), tableType);
+                    missingValueForNotNullableColumn = true;
+                }
+                else {
+                    expression = new Cast(new NullLiteral(), toSqlType(column.getType()));
+                }
             }
             else {
                 Symbol input = visibleFieldMappings.get(index);
-                Type tableType = column.getType();
                 Type queryType = symbolAllocator.getTypes().get(input);
 
                 if (queryType.equals(tableType) || typeCoercion.isTypeOnlyCoercion(queryType, tableType)) {
-                    assignments.put(output, input.toSymbolReference());
+                    expression = input.toSymbolReference();
                 }
                 else {
-                    Expression cast = noTruncationCast(input.toSymbolReference(), queryType, tableType);
-                    assignments.put(output, cast);
+                    expression = noTruncationCast(input.toSymbolReference(), queryType, tableType);
                 }
-                insertedColumnsBuilder.add(column);
             }
+            if (!column.isNullable() && !missingValueForNotNullableColumn) {
+                expression = new CoalesceExpression(expression, createNullNotAllowedFailExpression(column.getName(), tableType));
```

Comment:
why do we need to test `missingValueForNotNullableColumn` here?

Reply:
To make sure that we didn't already set it here:
`
if (!column.isNullable()) {
                    expression = createNullNotAllowedFailExpression(column.getName(), tableType);
                    missingValueForNotNullableColumn = true;
                }
`

Reply:
right, but the above could be simplified -- https://github.com/trinodb/trino/pull/13462#discussion_r941093005

Reply:
so I thought this way it is more efficient, it will fail on the first expression in the coercion, but sure I can change it